### PR TITLE
Improve `array_concat` signature for null and empty array

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #.
Part of #7142

### CODE REVIEW

File: datafusion/expr/src/signature.rs [hunk 1]
```diff
@@ -122,6 +122,9 @@ pub enum TypeSignature {
     /// List dimension of the List/LargeList is equivalent to the number of List.
     /// List dimension of the non-list is 0.
     ArrayAndElement,
+    /// Specialized Signature for ArrayConcat
+    /// Accept arbitrary arguments but they SHOULD be List/LargeList or Null, and the list dimension MAY NOT be the same.
+    ArrayConcat,
```
Comment (alamb) [inline datafusion/expr/src/signature.rs:127]: is there any usecase for this type of signature other than `ArrayConcat`? If there is only a single function that would likely have this signature, the code probably doesn't belong in the `TypeSignature` enum

I don't fully understand 

> and the list dimension MAY NOT be the same.


For example, one of the tests is

```sql
select array_concat([1, null], [null]);
```

Doesn't that have two arguments of the same dimension?

Comment (jayzhan211) [reply datafusion/expr/src/signature.rs:127]: In this case, where should be handle arrayConcat? I think there is no other array function share the same coercion with arrayConcat

Comment (jayzhan211) [reply datafusion/expr/src/signature.rs:127]: > ist dimension MAY NOT be the same.
I mean MAY or MAY NOT.
array concat is able to concat different dimension array in DF.

`select array_concat([1, null], [null]);` is the normal cases, dimensions are both 1.
`select array_concat([1, null], null);` is also valid, it acts like `array_append`, where dimensions are 1 and 0.

Comment (Weijun-H) [reply datafusion/expr/src/signature.rs:127]: When I worked on #8902, I thought this pr would benefit functions like `array_union`, and `array_except`. Their arguments are array and array.

Comment (jayzhan211) [reply datafusion/expr/src/signature.rs:127]: @Weijun-H But array_concat accepts list dimension with 1 and 3 (any n and m). union, intersect should be n and n-1

**Pull request closed without merging**